os/drivers/usbhost/uvc/usbhost_uvc_video.c

```diff
@@ -0,0 +1,1190 @@
+ * Name: uvc_set_video_ctrl
+ *
+ * Description:
+ *   Sets the video control options via the uvc_query_ctrl function.
+ *
+ * Input Parameters:
+ *   stream - reference to stream object
+ *   ctrl - reference to control object
+ *   probe - reference to probe object
+ *
+ * Returned Value:
+ *   OK on success, appropriate error on failures
+ ****************************************************************************/
+static inline int uvc_set_video_ctrl(struct uvc_streaming *stream, struct uvc_streaming_control *ctrl, int probe)
+{
+	uint16_t size = uvc_video_ctrl_size(stream);
+	uint8_t *data;
+	int ret;
+
+	data = kmm_zalloc(size);
+	if (data == NULL) {
+		uvdbg("Failed to allocate memory!!\n");
+		return -ENOMEM;
+	}
+
+	usbhost_uvc_putle16(&data[0], ctrl->bmHint);
+	data[2] = ctrl->bFormatIndex;
+	data[3] = ctrl->bFrameIndex;
+	usbhost_uvc_putle32(&data[4], ctrl->dwFrameInterval);
+	usbhost_uvc_putle16(&data[8], ctrl->wKeyFrameRate);
+	usbhost_uvc_putle16(&data[10], ctrl->wPFrameRate);
+	usbhost_uvc_putle16(&data[12], ctrl->wCompQuality);
+	usbhost_uvc_putle16(&data[14], ctrl->wCompWindowSize);
+	usbhost_uvc_putle16(&data[16], ctrl->wDelay);
+	usbhost_uvc_putle32(&data[18], ctrl->dwMaxVideoFrameSize);
+	usbhost_uvc_putle32(&data[22], ctrl->dwMaxPayloadTransferSize);
+
+	if (size >= 34) {
+		usbhost_uvc_putle32(&data[26], ctrl->dwClockFrequency);
+		data[30] = ctrl->bmFramingInfo;
+		data[31] = ctrl->bPreferedVersion;
+		data[32] = ctrl->bMinVersion;
+		data[33] = ctrl->bMaxVersion;
+	}
+
+	ret = uvc_query_ctrl(stream->dev, UVC_SET_CUR, 0, stream->intfnum, probe ? UVC_VS_PROBE_CONTROL : UVC_VS_COMMIT_CONTROL, data, size);
+	if (ret != size) {
+		uvdbg("Fail to query probe(%)\n", probe);
+		ret = -EIO;
+	} else {
+		ret = OK;
+	}
+
+	kmm_free(data);
+	return ret;
+}
+
+/****************************************************************************
+ * Name: uvc_get_video_ctrl
+ *
+ * Description:
+ *   This function retrieves the control settings for the specific control
+ *   option of the stream.
+ *
+ *   This function is called from the UVC_video methods.
+ *
+ * Input Parameters:
+ *   stream - uvc_streaming object
+ *   ctrl - UVC stream object control
+ *   probe - pointer to probe object
+ *   query - query type on stream object control
+ *
+ * Returned Values:
+ *   OK on success, appropriate error code on failure.
+ *
+ ****************************************************************************/
+int uvc_get_video_ctrl(struct uvc_streaming *stream, struct uvc_streaming_control *ctrl, int probe, uint8_t query)
+{
+	int ret = OK;
+	uint8_t *data;
+	struct uvc_format *format = NULL;
+	struct uvc_frame *frame = NULL;
+	uint16_t size = uvc_video_ctrl_size(stream);
+
+	data = kmm_zalloc(size);
+	if (data == NULL) {
+		uvdbg("Fail to allocate memory\n");
+		return -ENOMEM;
+	}
+
+	ret = uvc_query_ctrl(stream->dev, query, 0, stream->intfnum, probe ? UVC_VS_PROBE_CONTROL : UVC_VS_COMMIT_CONTROL, data, size);
```

Comment:
Can `stream` be NULL?

Reply:
stream object is allocated in function uvc_register_camera, if it fails there we will never reach in this function. Hence explicit check for stream is not required in all functions of this file.